lib/model/folder_sendrecv.go

```diff
@@ -900,36 +900,26 @@ func (f *sendReceiveFolder) renameFile(cur, source, target protocol.FileInfo, db
 	} else {
 		err = osutil.TryRename(f.fs, source.Name, tempName)
 	}
+	if err != nil {
+		return err
+	}
 
-	if err == nil {
-		blockStatsMut.Lock()
-		blockStats["total"] += len(target.Blocks)
-		blockStats["renamed"] += len(target.Blocks)
-		blockStatsMut.Unlock()
-
-		// The file was renamed, so we have handled both the necessary delete
-		// of the source and the creation of the target. Fix-up the metadata,
-		// update the local index of the target file and rename from temp to real name.
+	blockStatsMut.Lock()
+	blockStats["total"] += len(target.Blocks)
+	blockStats["renamed"] += len(target.Blocks)
+	blockStatsMut.Unlock()
 
-		dbUpdateChan <- dbUpdateJob{source, dbUpdateDeleteFile}
+	// The file was renamed, so we have handled both the necessary delete
+	// of the source and the creation of the target temp file. Fix-up the metadata,
+	// update the local index of the target file and rename from temp to real name.
 
-		if err = f.performFinish(nil, target, curTarget, true, tempName, dbUpdateChan, scanChan); err != nil {
-			return
-		}
-	} else {
-		// We failed the rename so we have a source file that we still need to
-		// get rid of. Attempt to delete it instead so that we make *some*
-		// progress. The target is unhandled.
+	if err = f.performFinish(nil, target, curTarget, true, tempName, dbUpdateChan, scanChan); err != nil {
+		return err
+	}
```

Comment:
This is a bit sketchy. The old file is gone the new file is there, yet chmod et al fails and db is not updated. As a minimum I suspect this will result in a conflict? What are the other failure scenarios?

Reply:
In the old case perform finish failing would imply success which I guess is also wrong.

Reply:
In the old case failure was not really handled at all. Now there shouldn't be any conflicts: If rename failed, nothing happened. If performFinish fails it is the same as with regular pulls: A temp file remains. As the del and create are then handled separately, the temp file can even be reused (even if most likely the same error will happen again on performFinish). The deletion isn't a problem either, it will just notice that the file is already gone and send the appropriate db update.

Reply:
A right, I forgot what performFinish does.